### PR TITLE
Query pipeline honors effective partition key property

### PR DESCRIPTION
**Pull request by j82w**
Query pipeline honors effective partition key property when deciding target ranges.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs [hunk 1]
```diff
@@ -132,6 +132,13 @@ internal static class DocumentQueryExecutionContextFactory
                                         collection.PartitionKey))
                             };
                         }
+                        else if (TryGetEpkProperty(feedOptions, out string effectivePartitionKeyString))
+                        {
+                            queryRanges = new List<Range<string>>
+                            {
+                                Range<string>.GetPointRange(effectivePartitionKeyString)
+                            };
+                        }
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs]: this control flow can be cleaned up a bit:

```
if feedOptions has physical partition key range id:
else if feedOptions has logical partition key:
else if feedOptions has effective partition key:
else 
  use query ranges from partition query execution info
```

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs [hunk 2]
```diff
@@ -288,5 +295,23 @@ private static bool IsOffsetLimitQuery(PartitionedQueryExecutionInfo partitioned
         {
             return partitionedQueryExecutionInfo.QueryInfo.HasOffset && partitionedQueryExecutionInfo.QueryInfo.HasLimit;
         }
+
+        private static bool TryGetEpkProperty(FeedOptions feedOptions, out string effectivePartitionKeyString)
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs]: use this same pattern for pkrangeid and logical partition key.

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs [hunk 3]
```diff
@@ -288,5 +295,23 @@ private static bool IsOffsetLimitQuery(PartitionedQueryExecutionInfo partitioned
...
+                effectivePartitionKeyString = effectivePartitionKeyStringObject as string;
+                if (string.IsNullOrEmpty(effectivePartitionKeyString))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(effectivePartitionKeyString));
+                }
+
+                return true;
+            }
+
+            effectivePartitionKeyString = null;
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs]: set the out variable to `default(T)` at the start.

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs [hunk 4]
```diff
@@ -288,5 +295,23 @@ private static bool IsOffsetLimitQuery(PartitionedQueryExecutionInfo partitioned
         {
             return partitionedQueryExecutionInfo.QueryInfo.HasOffset && partitionedQueryExecutionInfo.QueryInfo.HasLimit;
         }
+
+        private static bool TryGetEpkProperty(FeedOptions feedOptions, out string effectivePartitionKeyString)
+        {
+            if (feedOptions?.Properties != null && feedOptions.Properties.TryGetValue(
+                   WFConstants.BackendHeaders.EffectivePartitionKeyString, out object effectivePartitionKeyStringObject))
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs]: nit: each parameter on it's own line.

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs [hunk 5]
```diff
@@ -288,5 +295,23 @@ private static bool IsOffsetLimitQuery(PartitionedQueryExecutionInfo partitioned
...
+            {
+                effectivePartitionKeyString = effectivePartitionKeyStringObject as string;
+                if (string.IsNullOrEmpty(effectivePartitionKeyString))
+                {
+                    throw new ArgumentOutOfRangeException(nameof(effectivePartitionKeyString));
+                }
+
+                return true;
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs]: avoid early exits.

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs [hunk 6]
```diff
@@ -89,8 +89,8 @@ internal static class DocumentQueryExecutionContextFactory
 
             // If isContinuationExpected is false, we want to check if there are aggregates.
             if (
-                resourceTypeEnum.IsCollectionChild() 
-                && resourceTypeEnum.IsPartitioned() 
+                resourceTypeEnum.IsCollectionChild()
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs]: where are the tests for this change?

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs (hunk 5 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextFactory.cs]: just return if the out variable equals `default(T)` at the end.

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs [hunk 7]
```diff
@@ -334,16 +334,16 @@ public DocumentServiceRequest CreateDocumentServiceRequest(INameValueCollection
                 request,
                 cancellationToken);
 
-            return this.GetFeedResponse(request, documentServiceResponse);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs]: >this [](start = 19, length = 4)

Why?

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs [hunk 8]
```diff
@@ -425,6 +425,15 @@ public async Task<PartitionKeyRange> GetTargetPartitionKeyRangeById(string colle
             return range;
         }
 
+        public Task<List<PartitionKeyRange>> GetTargetPartitionKeyRangesByEpkString(string collectionResourceId, string effectivePartitionKeyString)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs]: >public [](start = 8, length = 6)

Of-course below is public but do they really need to be?

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs (hunk 7 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs]: My VS setting got reset. It should b e adding it. Not removing it. Will fix in next iteration.

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs (hunk 8 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs]: The class is internal.

File: Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs [hunk 9]
```diff
@@ -428,18 +428,27 @@ public async Task<PartitionKeyRange> GetTargetPartitionKeyRangeById(string colle
...
+        {
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs]: Please assert that EPK is not empty or NULL

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Query/DocumentQueryExecutionContextBase.cs]: Or mark this API private with assumption that caller does the validation

**Pull request merged**